performance-tests/bench_2/worker/WorkerDataReaderListener.h

```diff
@@ -36,6 +36,12 @@ class WorkerDataReaderListener : public Builder::DataReaderListener {
   void set_datareader(Builder::DataReader& datareader) override;
   void unset_datareader(Builder::DataReader& datareader) override;
 
+  size_t get_match_count() { return match_count_; }
+  size_t get_expected_match_count() { return expected_match_count_; }
+
+  std::condition_variable dr_expected_match_cv;
+  std::mutex dr_expected_match_cv_mutex;
```

Comment:
I think we should move the wait code inside the listener classes so we can make use of their internal mutex while we're checking match_count_ & expected_match_count_. So we wouldn't need the new mutex or the two new methods, but instead we'd add something like:

```
bool wait_for_expected_match(const std::chrono::system_clock::time_point& deadline) const {
  std::unique_lock<std::mutex> lock(mutex_);
  while (match_count_ != expected_match_count_) {
    if (wait_until(lock, timeout_time) == std::cv_status::timeout) {
      return match_count == expected_match_count_;
    }
  }
  return true;
}

```
Which would return true if the expected match was reached before deadline (otherwise false). Same for the writer listener. We'll need to make mutex_ mutable for the const to work, but that's fine. And then we just call those with our timeout and let the listeners worry about match counts etc.